### ISSUE TITLE
[Phase 6] Feedback Loop & Adaptation

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/high-dim-anomaly-detection#7]
## Summary

Implement the feedback loop for incorporating analyst decisions. This phase enables the system to learn from human feedback, reducing false positives and adapting to legitimate distribution shifts.

**Part of:** Meta Issue #1
**Dependencies:** Phase 3 (detection), Phase 5 (explanation)

## Objectives

1. Store analyst decisions in SQLite database
2. Adjust detection thresholds based on feedback
3. Trigger model retraining when drift detected
4. Provide feedback statistics and audit trail

## Files to Create

| File | Lines (est.) | Action | Description |
|------|-------------|--------|-------------|
| `src/feedback/__init__.py` | 1 | Create | Package marker |
| `src/feedback/decision_store.py` | 150 | Create | SQLite storage |
| `src/feedback/threshold_adapter.py` | 100 | Create | Threshold adjustment |
| `src/feedback/retrain_trigger.py` | 80 | Create | Retraining logic |
| `src/feedback/manager.py` | 150 | Create | Coordination |
| `tests/test_feedback.py` | 150 | Create | Feedback tests |

### Decision Store
- [ ] Create SQLite database with decisions table
- [ ] Store: sample_id, timestamp, score, prediction, analyst_label, is_correct
- [ ] Add indexes for efficient querying
- [ ] Implement `record()`, `get_recent()`, `get_statistics()`
- [ ] Handle database creation/migration

### Threshold Adapter
- [ ] Track false positive and false negative rates
- [ ] Adjust threshold with learning rate damping
- [ ] Require minimum samples before adjustment
- [ ] Store threshold history

### Retrain Trigger
- [ ] Count decisions since last training
- [ ] Compute drift metric (e.g., increasing FP rate)
- [ ] Signal retraining when thresholds exceeded
- [ ] Collect training samples for retraining

### Feedback Manager
- [ ] Coordinate all feedback components
- [ ] Expose simple API: `record_decision()`, `should_retrain()`
- [ ] Get feedback statistics
- [ ] Export feedback data for analysis

## Code Snippet: DecisionStore

```python
# src/feedback/decision_store.py
"""SQLite storage for analyst feedback."""

import sqlite3
from datetime import datetime
from pathlib import Path
from dataclasses import dataclass
from typing import List, Optional


@dataclass
class FeedbackDecision:
    """Single analyst feedback decision."""
    id: Optional[int]
    sample_id: str
    timestamp: datetime
    anomaly_score: float
    model_prediction: bool  # Model said anomaly?
    analyst_label: str      # 'anomaly', 'normal', 'unknown'
    is_correct: Optional[bool]
    notes: Optional[str]


class DecisionStore:
    """SQLite-based storage for analyst feedback."""

    def __init__(self, config: dict):
        storage_config = config.get('feedback', {}).get('storage', {})
        db_path = storage_config.get('db_path', 'data/feedback.db')

        self.db_path = Path(db_path)
        self.db_path.parent.mkdir(parents=True, exist_ok=True)
        self._init_database()

    def _init_database(self) -> None:
        """Create tables if needed."""
        with sqlite3.connect(self.db_path) as conn:
            conn.execute('''
                CREATE TABLE IF NOT EXISTS decisions (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    sample_id TEXT NOT NULL,
                    timestamp TEXT NOT NULL,
                    anomaly_score REAL NOT NULL,
                    model_prediction INTEGER NOT NULL,
                    analyst_label TEXT NOT NULL,
                    is_correct INTEGER,
                    notes TEXT
                )
            ''')
            conn.execute('CREATE INDEX IF NOT EXISTS idx_timestamp ON decisions(timestamp)')

    def record(self, decision: FeedbackDecision) -> int:
        """Store a decision, return inserted ID."""
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.execute('''
                INSERT INTO decisions
                (sample_id, timestamp, anomaly_score, model_prediction, analyst_label, is_correct, notes)
                VALUES (?, ?, ?, ?, ?, ?, ?)
            ''', (
                decision.sample_id,
                decision.timestamp.isoformat(),
                decision.anomaly_score,
                int(decision.model_prediction),
                decision.analyst_label,
                int(decision.is_correct) if decision.is_correct is not None else None,
                decision.notes
            ))
            return cursor.lastrowid

    def get_recent(self, limit: int = 100) -> List[FeedbackDecision]:
        """Get most recent decisions."""
        with sqlite3.connect(self.db_path) as conn:
            conn.row_factory = sqlite3.Row
            cursor = conn.execute(
                'SELECT * FROM decisions ORDER BY timestamp DESC LIMIT ?',
                (limit,)
            )
            return [self._row_to_decision(row) for row in cursor.fetchall()]

    def get_statistics(self) -> dict:
        """Get summary statistics."""
        with sqlite3.connect(self.db_path) as conn:
            total = conn.execute('SELECT COUNT(*) FROM decisions').fetchone()[0]
            correct = conn.execute('SELECT COUNT(*) FROM decisions WHERE is_correct = 1').fetchone()[0]
            labeled = conn.execute('SELECT COUNT(*) FROM decisions WHERE is_correct IS NOT NULL').fetchone()[0]

            return {
                'total_decisions': total,
                'accuracy': correct / labeled if labeled > 0 else None,
                'labeled_count': labeled
            }

    def _row_to_decision(self, row) -> FeedbackDecision:
        return FeedbackDecision(
            id=row['id'],
            sample_id=row['sample_id'],
            timestamp=datetime.fromisoformat(row['timestamp']),
            anomaly_score=row['anomaly_score'],
            model_prediction=bool(row['model_prediction']),
            analyst_label=row['analyst_label'],
            is_correct=bool(row['is_correct']) if row['is_correct'] is not None else None,
            notes=row['notes']
        )
```

## Code Snippet: ThresholdAdapter

```python
# src/feedback/threshold_adapter.py
"""Adaptive threshold adjustment based on feedback."""

from typing import List, Tuple
import numpy as np


class ThresholdAdapter:
    """Adjust anomaly threshold based on analyst feedback."""

    def __init__(self, config: dict):
        adapt_config = config.get('feedback', {}).get('adaptation', {})

        self.learning_rate = adapt_config.get('learning_rate', 0.1)
        self.min_samples = adapt_config.get('min_samples', 10)

        self._threshold = 0.5
        self._fp_history: List[Tuple[float, bool]] = []  # (score, was_fp)
        self._fn_history: List[Tuple[float, bool]] = []  # (score, was_fn)

    @property
    def current_threshold(self) -> float:
        return self._threshold

    def update(self, score: float, was_anomaly: bool, analyst_says_anomaly: bool) -> None:
        """
        Update threshold based on feedback.

        Args:
            score: Anomaly score of sample
            was_anomaly: Model predicted anomaly
            analyst_says_anomaly: Analyst labeled as anomaly
        """
        # Track false positives and negatives
        if was_anomaly and not analyst_says_anomaly:
            self._fp_history.append((score, True))
        elif not was_anomaly and analyst_says_anomaly:
            self._fn_history.append((score, True))

        # Only adjust if enough samples
        if len(self._fp_history) + len(self._fn_history) < self.min_samples:
            return

        # Compute adjustment
        fp_rate = len(self._fp_history) / (len(self._fp_history) + len(self._fn_history) + 1)

        # If too many FPs, raise threshold; if too many FNs, lower it
        if fp_rate > 0.5:
            # Too many false positives - be less sensitive
            adjustment = self.learning_rate * (fp_rate - 0.5)
            self._threshold = min(1.0, self._threshold + adjustment)
        else:
            # Too many false negatives - be more sensitive
            adjustment = self.learning_rate * (0.5 - fp_rate)
            self._threshold = max(0.0, self._threshold - adjustment)

    def reset_history(self) -> None:
        """Clear feedback history (e.g., after retraining)."""
        self._fp_history.clear()
        self._fn_history.clear()
```

## Code Snippet: FeedbackManager

```python
# src/feedback/manager.py
"""Coordinate all feedback components."""

from datetime import datetime
from typing import Optional
import numpy as np

from .decision_store import DecisionStore, FeedbackDecision
from .threshold_adapter import ThresholdAdapter
from .retrain_trigger import RetrainTrigger


class FeedbackManager:
    """Manage analyst feedback loop."""

    def __init__(self, config: dict):
        self.config = config
        self.store = DecisionStore(config)
        self.adapter = ThresholdAdapter(config)
        self.retrain_trigger = RetrainTrigger(config)

    def record_decision(
        self,
        sample_id: str,
        anomaly_score: float,
        model_said_anomaly: bool,
        analyst_label: str,  # 'anomaly', 'normal', 'unknown'
        notes: Optional[str] = None
    ) -> int:
        """
        Record analyst decision and update system.

        Returns:
            ID of recorded decision
        """
        # Determine correctness
        if analyst_label == 'unknown':
            is_correct = None
        else:
            analyst_says_anomaly = (analyst_label == 'anomaly')
            is_correct = (model_said_anomaly == analyst_says_anomaly)

        # Store decision
        decision = FeedbackDecision(
            id=None,
            sample_id=sample_id,
            timestamp=datetime.now(),
            anomaly_score=anomaly_score,
            model_prediction=model_said_anomaly,
            analyst_label=analyst_label,
            is_correct=is_correct,
            notes=notes
        )
        decision_id = self.store.record(decision)

        # Update threshold
        if analyst_label != 'unknown':
            self.adapter.update(
                score=anomaly_score,
                was_anomaly=model_said_anomaly,
                analyst_says_anomaly=(analyst_label == 'anomaly')
            )

        # Check for retraining
        self.retrain_trigger.record_decision(is_correct)

        return decision_id

    @property
    def current_threshold(self) -> float:
        return self.adapter.current_threshold

    def should_retrain(self) -> bool:
        return self.retrain_trigger.should_trigger()

    def get_statistics(self) -> dict:
        stats = self.store.get_statistics()
        stats['current_threshold'] = self.current_threshold
        stats['retrain_pending'] = self.should_retrain()
        return stats
```

## Verification

```python
from src.feedback.manager import FeedbackManager

config = {
    'feedback': {
        'storage': {'db_path': 'data/test_feedback.db'},
        'adaptation': {'learning_rate': 0.1, 'min_samples': 5},
        'retrain': {'trigger_threshold': 50}
    }
}

manager = FeedbackManager(config)
initial_threshold = manager.current_threshold

# Simulate false positives (model said anomaly, analyst said normal)
for i in range(10):
    manager.record_decision(
        sample_id=f"sample_{i}",
        anomaly_score=0.8,
        model_said_anomaly=True,
        analyst_label='normal'
    )

# Threshold should increase (be less sensitive)
print(f"Threshold: {initial_threshold:.3f} -> {manager.current_threshold:.3f}")
assert manager.current_threshold > initial_threshold, "Threshold should increase after FPs"

# Get statistics
stats = manager.get_statistics()
print(f"Statistics: {stats}")
```

## Definition of Done

- [ ] Decisions persist across sessions (SQLite)
- [ ] Threshold adapts correctly (increases after FPs, decreases after FNs)
- [ ] Retrain triggered after significant drift
- [ ] Statistics accurately reflect feedback history
- [ ] Unit tests pass

## Learning Notes

**Concepts to understand:**
- SQLite for embedded databases
- Online threshold adaptation
- Concept drift detection
- Dataclass for structured records
